lib/bibliothecary/parsers/npm.rb

```diff
@@ -210,6 +214,63 @@ def self.parse_v2_yarn_lock(contents, source = nil)
           end
       end
 
+      # This method currently has been tested to support:
+      #   lockfileVersion: '9.0'
+      #   lockfileVersion: '6.0'
+      #   lockfileVersion: '5.4'
+      def self.parse_pnpm_lock(contents, _source = nil)
+        parsed = YAML.load(contents)
+        lockfile_version = parsed["lockfileVersion"].to_i
+
+        dev_dependencies = parsed.dig("importers", ".", "devDependencies") # <= v9
+        dev_dependencies ||= parsed["devDependencies"] # <v9
+
+        # "dependencies" is in "packages" for < v9 and in "snapshots" for >= v9
+        # as of https://github.com/pnpm/pnpm/pull/7700.
+        (parsed["snapshots"] || parsed["packages"])
+          .map do |name_version, details|
+            name, version = case lockfile_version
+                            when 5
+                              # e.g. '/debug/2.6.9:'
+                              n, v = *name_version.sub(/^\//, "").split("/", 2)
+                              # e.g. '/debug/2.2.0_supports-color@1.2.0:'
+                              v = v.split("_", 2)[0]
+                            when 6
+                              # e.g. '/debug@2.6.9:'
+                              n, v = name_version.sub(/^\//, "").split("@", 2)
+                              # e.g. "debug@2.2.0(supports-color@1.2.0)"
+                              v = v.split("(", 2).first
+                            else
+                              # e.g. 'debug@2.6.9:'
+                              n, v = name_version.split("@", 2)
+                              # e.g. "debug@2.2.0(supports-color@1.2.0)"
+                              v = v.split("(", 2).first
+                            end
+            [n, v]
+
+            # TODO: the "dev" field was removed in v9 lockfiles (https://github.com/pnpm/pnpm/pull/7808)
+            # so this will only exist in v6 and below and might be unreliable.
+            # The proper way to set this for v9+ is to build a lookup of deps to
+            # their "dependencies", and then recurse through each package's
+            # parents. If the direct dep(s) that required them are all
+            # "devDependencies" then we can consider them "dev == true". This
+            # should be done using a DAG data structure, though, to be efficient
+            # and avoid cycles.
+            is_dev = details["dev"] == true
```

Comment:
re: this comment, I didn't have time to build a DAG thing, but we could revisit. Until then, "dev" will only be mostly-accurate for PNPM <9